Sources/Styleguide/DesignSystems/Spacing.swift

```diff
@@ -1,116 +1,139 @@
 import Css
 import Prelude
 
-extension Class {
-  public enum padding {
-    public static func all(_ n: Int) -> CssSelector {
-      return .class("p\(n)")
-    }
-
-    public static func top(_ n: Int) -> CssSelector {
-      return .class("pt\(n)")
-    }
-
-    public static func right(_ n: Int) -> CssSelector {
-      return .class("pr\(n)")
-    }
-
-    public static func bottom(_ n: Int) -> CssSelector {
-      return .class("pb\(n)")
-    }
-
-    public static func left(_ n: Int) -> CssSelector {
-      return .class("pl\(n)")
-    }
+public enum Side: String {
+  case all = ""
+  case bottom = "b"
+  case left = "l"
+  case leftRight = "x"
+  case right = "r"
+  case top = "t"
+  case topBottom = "y"
+}
 
-    public static func leftRight(_ n: Int) -> CssSelector {
-      return .class("px\(n)")
-    }
+extension Class {
+  public static func padding(_ data: [_Breakpoint: [Side: Int]]) -> CssSelector {
+    return selector(data, whitespace: .padding)
+  }
 
-    public static func topBottom(_ n: Int) -> CssSelector {
-      return .class("py\(n)")
-    }
+  public static func margin(_ data: [_Breakpoint: [Side: Int]]) -> CssSelector {
+    return selector(data, whitespace: .margin)
   }
+}
 
-  public enum margin {
-    public static func all(_ n: Int) -> CssSelector {
-      return .class("m\(n)")
-    }
+public let spacingStyles =
+  responsivePaddingStyles
+    <> responsiveMarginStyles
 
-    public static func top(_ n: Int) -> CssSelector {
-      return .class("mt\(n)")
-    }
+extension Side {
+  fileprivate static let allSides: [Side] = [.bottom, .left, .right, .top]
+}
 
-    public static func right(_ n: Int) -> CssSelector {
-      return .class("mr\(n)")
-    }
+private enum Whitespace: String {
+  case margin = "m"
+  case padding = "p"
 
-    public static func bottom(_ n: Int) -> CssSelector {
-      return .class("mb\(n)")
-    }
+  static let all: [Whitespace] = [.margin, .padding]
+}
 
-    public static func left(_ n: Int) -> CssSelector {
-      return .class("ml\(n)")
+private let responsivePaddingStyles =
```

Comment:
this wild 3-deep nested loop generates all the utility classes for padding on all sides, all breakpoints, and all values.

amazingly, the generated css is _smaller_ than the previous way of doing this (which didnt even include breakpoints).

Reply:
Cool!